### PR TITLE
Benja

### PR DESCRIPTION
**Pull request by jochman**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                        }),
+                    status=200,
+                    headers=headers
+                )
+    return err_resp
+
+
+def restart_jsons():
+    with open("./res/payments.json", "r+") as f2:
+        data = json.load(f2)
```
Comment (jochman) [inline app.py]: why load than overwrite the content?

File: app.py [hunk 2]
```diff
@@ -57,6 +61,7 @@ def post():
 
 
 @app.route(f"{api_path}/v1/consents/{consent}/status", methods=["GET"])
+@app.route(f"{api_path}/v1/consents/{consent}/status/", methods=["GET"])
```
Comment (jochman) [inline app.py]: Why this is doubled? think we don't need `/`

File: app.py [hunk 3]
```diff
@@ -17,15 +17,18 @@
...
 
 @app.route(api_path + "/v1/consents", methods=["POST"])
+@app.route(api_path + "/v1/consents/", methods=["POST"])
 def post():
-    global consent_created, current_req_id
+    global consent_created, current_req_id, payment_id
+    restart_jsons()
     x_req_id = request.headers.get("X-Request-ID")
     accounts = json.loads(request.data).get("access")
     # validating consent
```
Comment (UrielYair) [inline app.py:33]: why validating a consent if this is a consent request? 
(the goal of this endpoint is to request for a new consent for different resources in our DB: Accounts, Balances, Transactions)

File: app.py [hunk 4]
```diff
@@ -17,15 +17,18 @@
...
 }
 
 
 @app.route(api_path + "/v1/consents", methods=["POST"])
+@app.route(api_path + "/v1/consents/", methods=["POST"])
 def post():
-    global consent_created, current_req_id
+    global consent_created, current_req_id, payment_id
+    restart_jsons()
     x_req_id = request.headers.get("X-Request-ID")
```
Comment (UrielYair) [inline app.py:31]: missing information from JSON request body that needs to be considered:

- recurringIndicator : [Boolean]      // true, if the consent is for recurring
                                                            access to the account data
                                                      // false, if the consent is for one
                                                          access to the account data
-  validUntil : [ISODate]
-   frequencyPerDay : [Integer]         // maximum amount of access per day.

File: app.py [hunk 5]
```diff
@@ -35,6 +38,7 @@ def post():
             return err_resp
     consent_created = True
     current_req_id = x_req_id
+    payment_id = 0
     return Response(
```
Comment (UrielYair) [inline app.py:41]: The response will only return _links.
- scaRedirect : [value of TPP-Redirect-URI] - you will get it as a parameter in request header.
- self : [ link to: /v1/consents/{generatedConsentID} ]
- status : [string]  // Accepted/Rejected - more info in page 241.
- scaStatus : [string] // result of user approval checking. (we should simulate approval of the user to give consent to the TPP)

Comment (UrielYair) [reply app.py:41]: status is good as you wrote.
what I meant is that we should make sure that the result will be stored in this link.

-same for scaStatus

File: app.py [hunk 6]
```diff
@@ -17,15 +17,18 @@
...
 
 
 @app.route(api_path + "/v1/consents", methods=["POST"])
+@app.route(api_path + "/v1/consents/", methods=["POST"])
 def post():
-    global consent_created, current_req_id
+    global consent_created, current_req_id, payment_id
+    restart_jsons()
     x_req_id = request.headers.get("X-Request-ID")
     accounts = json.loads(request.data).get("access")
```
Comment (UrielYair) [inline app.py:32]: we should simulate user (PSU) approval for each iban that inside "access" ( access contain lists of: Accounts, Balances, Transactions. All of those are ibans values.

For examples:
"access" : {
       "Accounts" : [  
                                "iban" :  "~", ..... 
                            ],
       "Balances" : [  
                                "iban" :  "~", ..... 
                            ], 
       "Transactions" : [  
                                "iban" :  "~", ..... 
                            ]
}

Comment (UrielYair) [reply app.py:32]: I don't know why is the line across the words, read as usual and ignore these lines.

**Review by benjika — APPROVED** [review]

File: app.py [hunk 7]
```diff
@@ -103,6 +109,7 @@ def get_authorization():
 
 
 @app.route(api_path + "/v1/accounts", methods=["GET"])
+@app.route(api_path + "/v1/accounts/", methods=["GET"])
 def get_accounts():
     header = request.headers
     if header.get("X-Request-ID") == current_req_id and header.get("Consent-ID") == consent:
```
Comment (UrielYair) [inline app.py]: from accounts - 
- delete:  product, cashAccountType

File: app.py [hunk 8]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+            return Response(
+                json.dumps(
+                    {
+                        "accounts":
+                            [
+                                {
+                                    "resourceId": "3dc3d5b3-7023-4848-9853-f5400a64e80f",
+                                    "iban": "LV95LATB000222PSD2001",
+                                    "currency": "EUR",
+                                    "product": "Girokonto",
```
Comment (UrielYair) [inline app.py:161]: delete product

File: app.py [hunk 9]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                                    "cashAccountType": "CurrentAccount",
```
Comment (UrielYair) [inline app.py:162]: delete cashAccountType.

File: app.py [hunk 10]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+            return Response(
+                json.dumps(
+                    {
+                        "accounts":
+                            [
+                                {
+                                    "resourceId": "3dc3d5b3-7023-4848-9853-f5400a64e81g",
+                                    "iban": "LV68LATB000222PSD2002",
+                                    "currency": "EUR",
+                                    "product": "Fremdwährungskonto",
```
Comment (UrielYair) [inline app.py:186]: delete product.

File: app.py [hunk 11]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                                    "cashAccountType": "CurrentAccount",
```
Comment (UrielYair) [inline app.py:187]: delete cashAccuntType

File: app.py [hunk 12]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+
+
+@app.route(api_path + "/v1/funds-confirmations", methods=["POST"])
+@app.route(api_path + "/v1/funds-confirmations/", methods=["POST"])
+def get_funds_confirmations():
+    header = request.headers
+    if header.get("X-Request-ID") == current_req_id and header.get("Consent-ID") == consent:
+        headers["X-Request-ID"] = current_req_id
+        headers["Content-Type"] = "application/json"
+        access = json.loads(request.data).get("access")
```
Comment (UrielYair) [inline app.py]: There is no access field, maybe you are confusing it with "account".

File: app.py [hunk 13]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+        req_iban = access["iban"]
```
Comment (UrielYair) [inline app.py]: same: change "access" to "account".

File: app.py [hunk 14]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                headers=headers
+            )
+    return err_resp
+
+
+@app.route(api_path + "/v1/payments/sepa-credit-transfers", methods=["POST"])
+@app.route(api_path + "/v1/payments/sepa-credit-transfers/", methods=["POST"])
+def transfer_sepa_credit():
+    header = request.headers
+    if header.get("X-Request-ID") == current_req_id and header.get("Consent-ID") == consent \
```
Comment (UrielYair) [inline app.py:253]: How does this if statement checks that the TPP who send this request has consent to the debtorAccount and the creditorAccount.

File: app.py [hunk 15]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                headers["TPP-Redirect-URI"] = tpp_redirect_uri
+                headers["ASPSP-SCA-Approach"] = "REDIRECT"
+                temp_payment_id = payment_id - 1
+                return Response(
+                    json.dumps(
+                        {
+                            "paymentId": temp_payment_id,
+                            "_links": {
+                                "scaRedirect": tpp_redirect_uri,
+                                "self": "/v1/payments/" + str(temp_payment_id),
```
Comment (UrielYair) [inline app.py:284]: is this resource will now have a file or an entry with information regarding this payment transfer?

File: app.py [hunk 16]
```diff
@@ -143,9 +150,211 @@ def get_accounts():
...
+                                "status": "Rejected",
+                                "scaStatus": "Accepted"
+                            }
+                        }),
+                    status=200,
+                    headers=headers
+                )
+            else:
+                temp_payment_id = payment_id - 1
```
Comment (UrielYair) [inline app.py]: why using minus?

**Review by UrielYair — COMMENTED** [review]
please read.

File: app.py (hunk 4 above)
Comment (jochman) [reply app.py:31]: throw away recurring indicator

File: app.py (hunk 6 above)
Comment (jochman) [reply app.py:32]: because it is in MarkDown format.

File: app.py (hunk 3 above)
Comment (jochman) [reply app.py:33]: validating that the data we got is equal to what we expected.

File: app.py (hunk 5 above)
Comment (jochman) [reply app.py:41]: https://developer.pnbbanka.eu/en/docs/index?group=1#endpoint-27

File: app.py (hunk 6 above)
Comment (jochman) [reply app.py:32]: and we have only one flow that we determine

File: app.py (hunk 14 above)
Comment (jochman) [reply app.py:253]: no

**Pull request merged**